uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/contract-entitlements.js

```diff
@@ -0,0 +1,106 @@
+/* eslint-env amd */
+
+define([
+  'common/lodash',
+  'common/moment',
+  'leave-absences/absence-tab/modules/components',
+  'common/models/contract'
+], function (_, moment, components) {
+  components.component('contractEntitlements', {
+    bindings: {
+      contactId: '<'
+    },
+    templateUrl: ['settings', function (settings) {
+      return settings.pathTpl + 'components/contract-entitlements.html';
+    }],
+    controllerAs: 'entitlements',
+    controller: ['$log', '$q', 'HR_settings', 'AbsenceType', 'Contract', controller]
+  });
+
+  function controller ($log, $q, HRSettings, AbsenceType, Contract) {
+    $log.debug('Component: contract-entitlements');
+
+    var vm = {};
+
+    vm.contactId = this.contactId;
+    vm.contracts = [];
+    vm.contractsLoaded = false;
+    vm.absenceTypes = [];
+
+    (function init () {
+      return $q.all([
+        loadAbsenceTypes()
+      ])
+      .then(function () {
+        return loadContracts();
+      });
+    })();
+
+    /**
+     * Formats the date according to user settings
+     *
+     * @param {object} date
+     */
+    function formatDate (date) {
+      var userDateFormat = HRSettings.DATE_FORMAT;
+      var dateFormat = userDateFormat ? userDateFormat.toUpperCase() : 'DD/MM/YYYY';
+
+      return date ? moment(date).format(dateFormat) : null;
+    }
+
+    /**
+     * Loads absence types
+     *
+     * @return {Promise}
+     */
+    function loadAbsenceTypes () {
+      return AbsenceType.all()
+        .then(function (absenceTypes) {
+          vm.absenceTypes = absenceTypes;
+        });
+    }
+
+    /**
+     * Loads contracts
+     *
+     * @return {Promise}
+     */
+    function loadContracts () {
+      return Contract.all({contact_id: vm.contactId})
+        .then(function (data) {
+          setContractsProps(data);
+        });
+    }
+
+    /**
+     * Processes contracts from data and sets them to a controller
+     *
+     * @param {object} contracts
+     */
+    function setContractsProps (contracts) {
+      vm.contracts = _.sortBy(contracts, function (contract) {
+        return moment(contract.info.details.period_start_date);
+      }).map(function (contract) {
+        var info = contract.info;
+        var details = info.details;
+        var absences = _.map(vm.absenceTypes, function (absenceType) {
+          var leave = _.filter(info.leave, function (leave) {
+            return leave.leave_type === absenceType.id;
+          })[0];
+          return {
+            amount: leave ? leave.leave_amount : ''
+          };
+        });
+        return {
+          'position': details.position,
+          'start_date': formatDate(details.period_start_date),
+          'end_date': formatDate(details.period_end_date),
+          'absences': absences
+        };
+      });
+      vm.contractsLoaded = true;
```

Comment:
We had `vm.contracts = null;` which allowed us to understand if contracts we not loaded yet. Because we removed that the controller now needs to know if the contracts are still loading OR loaded and empty OR loaded and not empty. Because now `vm.contracts = [] // always array`, we need a third variable.